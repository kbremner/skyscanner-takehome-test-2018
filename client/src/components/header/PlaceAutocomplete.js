import React, { Component } from 'react';
import BpkAutosuggest, { BpkAutosuggestSuggestion } from 'bpk-component-autosuggest';
import PropTypes from '../../types';

const getSuggestionValue = ({ PlaceId }) => PlaceId.replace('-sky', '');

const renderSuggestion = ({ PlaceId, PlaceName, CountryName }) => (
  <BpkAutosuggestSuggestion
    value={`${PlaceName} (${PlaceId.replace('-sky', '')})`}
    subHeading={CountryName}
    tertiaryLabel="Airport"
  />
);

class PlaceAutocomplete extends Component {
  constructor(props) {
    super(props);

    this.state = {
      value: props.defaultValue
    };
  }

  onChange = (e, { newValue }) => {
    this.setState({
      value: newValue,
    });
  }

  onSuggestionsFetchRequested = ({ value }) => {
    this.props.onSearchForPlaces(value);
  }

  onSuggestionsClearRequested = () => {
    this.props.onClearPlaces();
  }

  onSuggestionSelected = (event, { suggestion }) => {
    this.props.onPlaceSelected(suggestion.PlaceId);
  }

  render() {
    const { value } = this.state;
    const { name, places } = this.props;

    const inputProps = {
      id: `place-search-${name}`,
      name,
      placeholder: 'Enter a place',
      value,
      onChange: this.onChange,
    };

    return <BpkAutosuggest
      suggestions={places}
      onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
      onSuggestionsClearRequested={this.onSuggestionsClearRequested}
      getSuggestionValue={getSuggestionValue}
      renderSuggestion={renderSuggestion}
      onSuggestionSelected={this.onSuggestionSelected}
      inputProps={inputProps}
    />
  }
}

PlaceAutocomplete.propTypes = {
  defaultValue: PropTypes.string,
  name: PropTypes.string.isRequired,
  places: PropTypes.arrayOf(PropTypes.place).isRequired,
  onSearchForPlaces: PropTypes.func.isRequired,
  onClearPlaces: PropTypes.func.isRequired,
  onPlaceSelected: PropTypes.func.isRequired
};

export default PlaceAutocomplete;
