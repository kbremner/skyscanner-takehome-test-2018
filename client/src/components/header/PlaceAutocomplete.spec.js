import React from 'react';
import { shallow } from 'enzyme';
import PlaceAutocomplete from './PlaceAutocomplete';

const createPlaces = size => Array.apply(null, Array(size))
  .map((val, i) => ({
    PlaceId: `test-place-id-${i}`,
    PlaceName: `test-place-name-${i}`,
    CountryName: `test-country-name-${i}`
  }));

const defaultProps = () => ({
  name: 'test-place-autocomplete',
  places: [],
  onSearchForPlaces: jest.fn(),
  onClearPlaces: jest.fn(),
  onPlaceSelected: jest.fn()
});

const setup = (overrideProps = {}) => {
  const props = {
    ...defaultProps(),
    ...overrideProps
  };
  const wrapper = shallow(<PlaceAutocomplete {...props} />);
  return { wrapper, props };
}

describe('PlaceAutocomplete', () => {
  it('renders', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with default value', () => {
    const { wrapper } = setup({ defaultValue: 'test-default' });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with results', () => {
    const { wrapper } = setup({ places: createPlaces(2) });
    expect(wrapper).toMatchSnapshot();
  });

  it('calls handler when suggestions fetch requested', () => {
    const value = Symbol('test-value');
    const { wrapper, props } = setup();
    const { onSearchForPlaces } = props;
    wrapper.find('Autosuggest').simulate('suggestionsFetchRequested', { value });
    expect(onSearchForPlaces).toHaveBeenCalledTimes(1);
    expect(onSearchForPlaces).toHaveBeenCalledWith(value);
  });

  it('calls handler when clearing of suggestions is requested', () => {
    const { wrapper, props } = setup();
    const { onClearPlaces } = props;
    wrapper.find('Autosuggest').simulate('suggestionsClearRequested');
    expect(onClearPlaces).toHaveBeenCalledTimes(1);
  });

  it('calls handler when suggestion selected', () => {
    const suggestion = { PlaceId: Symbol('test-place-id') };
    const { wrapper, props } = setup();
    const { onPlaceSelected } = props;
    wrapper.find('Autosuggest').simulate('suggestionSelected', null, { suggestion });
    expect(onPlaceSelected).toHaveBeenCalledTimes(1);
    expect(onPlaceSelected).toHaveBeenCalledWith(suggestion.PlaceId);
  });
});
