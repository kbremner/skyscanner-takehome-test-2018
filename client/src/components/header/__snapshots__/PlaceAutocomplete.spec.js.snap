// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceAutocomplete renders 1`] = `
<Autosuggest
  alwaysRenderSuggestions={false}
  focusInputOnSuggestionClick={true}
  getSuggestionValue={[Function]}
  highlightFirstSuggestion={false}
  id="1"
  inputProps={
    Object {
      "id": "place-search-test-place-autocomplete",
      "name": "test-place-autocomplete",
      "onChange": [Function],
      "placeholder": "Enter a place",
      "value": undefined,
    }
  }
  multiSection={false}
  onSuggestionSelected={[Function]}
  onSuggestionsClearRequested={[Function]}
  onSuggestionsFetchRequested={[Function]}
  renderInputComponent={[Function]}
  renderSuggestion={[Function]}
  renderSuggestionsContainer={[Function]}
  shouldRenderSuggestions={[Function]}
  suggestions={Array []}
  theme={
    Object {
      "container": "bpk-autosuggest__container",
      "containerOpen": "bpk-autosuggest__container--open",
      "sectionContainer": "bpk-autosuggest__section-container",
      "sectionTitle": "bpk-autosuggest__section-title",
      "suggestion": "bpk-autosuggest__suggestion-item",
      "suggestionHighlighted": "bpk-autosuggest__suggestion-item--highlighted",
      "suggestionsContainer": "bpk-autosuggest__suggestions-container",
      "suggestionsContainerOpen": "bpk-autosuggest__suggestions-container--open",
      "suggestionsList": "bpk-autosuggest__suggestions-list",
    }
  }
/>
`;

exports[`PlaceAutocomplete renders with default value 1`] = `
<Autosuggest
  alwaysRenderSuggestions={false}
  focusInputOnSuggestionClick={true}
  getSuggestionValue={[Function]}
  highlightFirstSuggestion={false}
  id="1"
  inputProps={
    Object {
      "id": "place-search-test-place-autocomplete",
      "name": "test-place-autocomplete",
      "onChange": [Function],
      "placeholder": "Enter a place",
      "value": "test-default",
    }
  }
  multiSection={false}
  onSuggestionSelected={[Function]}
  onSuggestionsClearRequested={[Function]}
  onSuggestionsFetchRequested={[Function]}
  renderInputComponent={[Function]}
  renderSuggestion={[Function]}
  renderSuggestionsContainer={[Function]}
  shouldRenderSuggestions={[Function]}
  suggestions={Array []}
  theme={
    Object {
      "container": "bpk-autosuggest__container",
      "containerOpen": "bpk-autosuggest__container--open",
      "sectionContainer": "bpk-autosuggest__section-container",
      "sectionTitle": "bpk-autosuggest__section-title",
      "suggestion": "bpk-autosuggest__suggestion-item",
      "suggestionHighlighted": "bpk-autosuggest__suggestion-item--highlighted",
      "suggestionsContainer": "bpk-autosuggest__suggestions-container",
      "suggestionsContainerOpen": "bpk-autosuggest__suggestions-container--open",
      "suggestionsList": "bpk-autosuggest__suggestions-list",
    }
  }
/>
`;

exports[`PlaceAutocomplete renders with results 1`] = `
<Autosuggest
  alwaysRenderSuggestions={false}
  focusInputOnSuggestionClick={true}
  getSuggestionValue={[Function]}
  highlightFirstSuggestion={false}
  id="1"
  inputProps={
    Object {
      "id": "place-search-test-place-autocomplete",
      "name": "test-place-autocomplete",
      "onChange": [Function],
      "placeholder": "Enter a place",
      "value": undefined,
    }
  }
  multiSection={false}
  onSuggestionSelected={[Function]}
  onSuggestionsClearRequested={[Function]}
  onSuggestionsFetchRequested={[Function]}
  renderInputComponent={[Function]}
  renderSuggestion={[Function]}
  renderSuggestionsContainer={[Function]}
  shouldRenderSuggestions={[Function]}
  suggestions={
    Array [
      Object {
        "CountryName": "test-country-name-0",
        "PlaceId": "test-place-id-0",
        "PlaceName": "test-place-name-0",
      },
      Object {
        "CountryName": "test-country-name-1",
        "PlaceId": "test-place-id-1",
        "PlaceName": "test-place-name-1",
      },
    ]
  }
  theme={
    Object {
      "container": "bpk-autosuggest__container",
      "containerOpen": "bpk-autosuggest__container--open",
      "sectionContainer": "bpk-autosuggest__section-container",
      "sectionTitle": "bpk-autosuggest__section-title",
      "suggestion": "bpk-autosuggest__suggestion-item",
      "suggestionHighlighted": "bpk-autosuggest__suggestion-item--highlighted",
      "suggestionsContainer": "bpk-autosuggest__suggestions-container",
      "suggestionsContainerOpen": "bpk-autosuggest__suggestions-container--open",
      "suggestionsList": "bpk-autosuggest__suggestions-list",
    }
  }
/>
`;
